pkg/ui/server.go

```diff
@@ -0,0 +1,282 @@
+func NewServer(ctx context.Context, cfg *ServiceConfig, tmplLocations map[string]string) (*Server, error) {
+	templateMap := make(map[string]*template.Template)
+
+	// Parse templates
+	for key, path := range tmplLocations {
+		tmpl, err := template.ParseFiles(path)
+		if err != nil {
+			return nil, fmt.Errorf("parsing %s template: %w", path, err)
+		}
+		templateMap[key] = tmpl
+	}
+
+	return &Server{
+		templates: templateMap,
+		allowList: cfg.AllowList,
+	}, nil
+}
+
+// Routes creates a ServeMux of all of the routes that
+// this Router supports.
+func (s *Server) Routes() http.Handler {
+	mux := http.NewServeMux()
+	fs := http.FileServer(http.Dir("./assets/static"))
+	mux.Handle("/assets/static/", http.StripPrefix("/assets/static/", fs))
+	mux.Handle("/popup", s.handlePopup())
+	return mux
+}
+
+func (s *Server) handlePopup() http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		handlePopupFunc(w, r, s.templates, s.allowList)
+	})
+}
+
+// handlePopupFunc the initial page load of the form as well as responds to the submission of the form and ultimately renders an HTML page.
+func handlePopupFunc(w http.ResponseWriter, r *http.Request, templates map[string]*template.Template, allowList []string) {
+	formDetails := getFormDetails(r)
+
+	// Initial page load, just render the page
+	if r.Method == "GET" {
+		// set some defaults for the form
+		formDetails.Category = categories[0]
+		formDetails.TTL = ttls[0]
+		render(w, templates["popup"], formDetails)
+		return
+	}
+
+	// Form submission
+	if r.Method == "POST" {
+		// 1. Check if the origin is part of the allowlist
+		if !validateOrigin(r.FormValue("origin"), allowList) {
+			forbiddenDetails := &ForbiddenDetails{
+				PageTitle: "JVS - Error page",
+				Message:   "Something went wrong",
+			}
+			render(w, templates["forbidden"], forbiddenDetails)
+			return
+		}
+
+		// 2. Validate input
+		if !validateForm(formDetails) {
+			render(w, templates["popup"], formDetails)
+			return
+		}
+
+		// 3. [TODO] Request a token
+		token := "token_from_server"
+
+		// 4. Redirect to a confirmation page with context, ultimately needed to postMessage back to the client
+		successDetails := &SuccessDetails{
+			PageTitle:    "JVS - Successful token retrieval",
+			Token:        token,
+			TargetOrigin: formDetails.Origin,
+			WindowName:   formDetails.WindowName,
+		}
+
+		render(w, templates["success"], successDetails)
+	}
+}
+
+// Checks the origin parameter against all entries in the allow list.
+func validateOrigin(originParam string, allowList []string) bool {
+	// origin: shop.acme.com
+	// allowList: [acme.com,*]
+
+	// Check if origin is localhost or private ip
+	if strings.HasPrefix(originParam, "http://localhost") {
+		return true
+	}
+
+	originSplit := strings.Split(originParam, ".")
+
+	for _, domain := range allowList {
+		// special case, allow everything
+		if domain == "*" {
+			return true
+		}
```

Comment:
If the allowed list include `*`, we should _only_ allow the list to contain a `*`. Otherwise `[acme.com, *]` doesn't make sense and is probably a bug.